src/python/grpcio_tests/tests/interop/xds_interop_client.py

```diff
@@ -0,0 +1,254 @@
+        end = None
+        watcher = None
+        with _global_lock:
+            start = _global_rpc_id + 1
+            end = start + request.num_rpcs
+            watcher = _StatsWatcher(start, end)
+            _watchers.add(watcher)
+        response = watcher.await_rpc_stats_response(request.timeout_sec)
+        with _global_lock:
+            _watchers.remove(watcher)
+        logger.info("Returning stats response: {}".format(response))
+        return response
+
+
+def _start_rpc(request_id: int, stub: test_pb2_grpc.TestServiceStub,
+               timeout: float, futures: Mapping[int, grpc.Future]) -> None:
+    logger.info(f"Sending request to backend: {request_id}")
+    future = stub.UnaryCall.future(messages_pb2.SimpleRequest(),
+                                   timeout=timeout)
+    futures[request_id] = future
+
+
+def _on_rpc_done(rpc_id: int, future: grpc.Future,
+                 print_response: bool) -> None:
+    exception = future.exception()
+    hostname = ""
+    if exception is not None:
+        if exception.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
+            logger.error(f"RPC {rpc_id} timed out")
+        else:
+            logger.error(exception)
+    else:
+        response = future.result()
+        logger.info(f"Got result {rpc_id}")
+        hostname = response.hostname
+        if print_response:
+            if future.code() == grpc.StatusCode.OK:
+                logger.info("Successful response.")
+            else:
+                logger.info(f"RPC failed: {call}")
+    with _global_lock:
+        for watcher in _watchers:
+            watcher.on_rpc_complete(rpc_id, hostname)
+
+
+def _remove_completed_rpcs(futures: Mapping[int, grpc.Future],
+                           print_response: bool) -> None:
+    logger.debug("Removing completed RPCs")
+    done = []
+    for future_id, future in futures.items():
+        if future.done():
+            _on_rpc_done(future_id, future, args.print_response)
+            done.append(future_id)
+    for rpc_id in done:
+        del futures[rpc_id]
+
+
+def _cancel_all_rpcs(futures: Mapping[int, grpc.Future]) -> None:
+    logger.info("Cancelling all remaining RPCs")
+    for future in futures.values():
+        future.cancel()
+
+
+def _run_single_channel(args: argparse.Namespace):
+    global _global_rpc_id  # pylint: disable=global-statement
+    duration_per_query = 1.0 / float(args.qps)
+    with grpc.insecure_channel(args.server) as channel:
+        stub = test_pb2_grpc.TestServiceStub(channel)
+        futures: Dict[int, grpc.Future] = {}
+        while not _stop_event.is_set():
+            request_id = None
+            with _global_lock:
+                request_id = _global_rpc_id
+                _global_rpc_id += 1
+            start = time.time()
+            end = start + duration_per_query
+            _start_rpc(request_id, stub, float(args.rpc_timeout_sec), futures)
+            _remove_completed_rpcs(futures, args.print_response)
```

Comment:
optional: Sweep and remove works. An alternative can be a factory method and `add_done_callback`, which calls callback automatically and we can focus on the scheduling part.

Reply:
I gave this a shot, but found that the callback isn't happening in the case of timeouts. Looks like a bug. Considering the difference is cosmetic, I'm going to stick with the sweep approach for this PR and fix [the bug](https://github.com/grpc/grpc/issues/22387) in a separate PR.